### ISSUE TITLE
Access-Control-Allow-Origin Error

### ISSUE THREAD

**Issue by marcel-goldschen-ohm** [maxlath/wikibase-sdk#66]
First, thanks so much for this wonderful interface to wikibase. Also, this may be an issue only with my understanding rather than an actual issue with wikibase-sdk. Regardless, hopefully I'll be set straight here.

I'm getting what seems like some odd behavior. I'm making a series of sparql queries as follows, where rp is request-promise:

async function getItems(Qid) {
  const sparql = `... ${Qid} ...`; // I've omitted my actual query here
  const url = wbk.sparqlQuery(sparql);
  return rp(url).then(wbk.simplify.sparqlResults);
}

results = await getItems(...);

All of this seems to work just fine at the beginning. I get the expected results for my sparql queries, life is good. However, for some queries with certain Qids I get the following error:

Origin http://localhost:... is not allowed by Access-Control-Allow-Origin.
Fetch API cannot load https://query.wikidata.org/sparql?format=json&query=... due to access control checks.

The same sparql works for another Qid. Any idea what is going on? Why does CORS seem to be a problem sometimes and not others? Very confused. I saw somewhere else that I should make sure I am requesting JSON, but the url has format=json so this should be fine I assume. Apologies in advance if this is something more fundamental that has nothing to do with wikibase-sdk. Any advice or pointers in the right direction will be greatly appreciated.

Cheers,
Marcel

**Comment by marcel-goldschen-ohm** [comment]
I've made a [small example](https://github.com/marcel-goldschen-ohm/test) that produces this error.

**Comment by maxlath** [comment]
hi!

I tried to run your example and might have a clue:
* It seems that your code generates a lot of requests, which trigger some `429 Too Many Requests` responses from query.wikidata.org (see [query limits in manual](https://www.mediawiki.org/wiki/Wikidata_Query_Service/User_Manual#Query_limits)).
* As your requests don't seem to generate errors, the 429 errors are either because:
  * you break the `One client is allowed 60 seconds of processing time each 60 seconds` rule
  * you run more than 5 requests in parallel (likely as you don't `await` for your nested call to `getItemTree`)
* 429 being an error code, it doesn't get the same headers as the other requests that get a 200 response; and in particular, it doesn't get the CORS headers. (the reason for this difference of behavior could be that the Nginx proxy between your browser and the Query Service doesn't apply the same headers for error responses

today I learned:
* that, while never seen before for me, `$item` is a valid SPARQL variable ^^
